### PR TITLE
Use npm to handle frontend dependencies

### PR DESCRIPTION
**Pull request by marudor**
Now with 50% more memory consumption! (in dev)
Also another dependency management tool! \o/

On a serious note:
composer isn't really up to date with dependencies for frontend stuff. NPM obviously is.
Also this modernises some of the code. And also handles old browsers better (babel preset env)
So older browser work and we dont have to care! \o/

### CODE REVIEW

File: js/moment-countdown.js [hunk 1]
```diff
@@ -1,8 +1,3 @@
-/**
- * Initialize all moment countdowns on the page. A moment countdown has the
- * class "moment-countdown" and the attribute "data-timestamp" which defines the
- * countdown's time goal.
- */
```
Comment (MyIgel) [inline js/moment-countdown.js]: I'm missing the comment ;)

File: package.json [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  "license": "MIT",
+  "scripts": {
+    "build": "NODE_ENV=production webpack",
+    "build:watch": "webpack --watch"
+  },
+  "dependencies": {
+    "bootstrap": "^3.3.7",
+    "bootstrap-datepicker": "^1.7.1",
+    "chart.js": "^1.0.2",
+    "icomoon": "^1.0.0",
```
Comment (MyIgel) [inline package.json]: Currently the icomoon font is a stripped down version that only contains the angel logo, that would load the whole font?
For reference: #426 (Feature: Logo and text should be configurable)

File: webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+const plugins = [
+  new webpack.DefinePlugin({
+    'process.env': {
+      NODE_ENV: JSON.stringify(nodeEnv),
+    },
+  }),
+  new ExtractTextPlugin('[name].css'),
+];
+
+// if (!__DEV__) {
```
Comment (MyIgel) [inline webpack.config.js]: Commented out code?

File: composer.json [hunk 4]
```diff
@@ -19,8 +19,7 @@
         "illuminate/container": "5.5.*",
         "psr/container": "^1.0",
         "psr/log": "^1.0",
-        "symfony/http-foundation": "^3.3",
-        "twbs/bootstrap": "^3.3"
```
Comment (MyIgel) [inline composer.json]: ❤️

File: .babelrc [hunk 5]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (MyIgel) [inline .babelrc]: Moar files in the root folder 🎉 or 😥?

File: webpack.config.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
+const path = require('path');
```
Comment (MyIgel) [inline webpack.config.js]: Webpack should be documented in the readme ;)

**Comment by msquare** [comment]
How do I use this? I guess there are more steps needed in README to make the system run?

**Comment by msquare** [comment]
@marudor @MyIgel What about this pr? Do you want to finish it for merging?

File: README.md [hunk 7]
```diff
@@ -13,20 +13,29 @@ To report bugs use [engelsystem/issues](https://github.com/engelsystem/engelsyst
...
  * Install project dependencies:
      ```bash
      composer install
+     yarn
      ```
     On production systems it is recommended to use
     ```bash
     composer install --no-dev
     composer dump-autoload --optimize
+
```
Comment (MyIgel) [inline README.md]: ?

**Comment by marudor** [comment]
As I'm not familiar with gitlab ci + we are on github here!
So no Idea. Please adjust it yourself based on the README instructions.

**Comment by MyIgel** [comment]
@marudor: After installing and building the project I get the following error in the console:

```
vendor.js:27659 Uncaught TypeError: Cannot create property 'exports' on boolean 'false'
    at vendor.js:27659
    at define (vendor.js:17367)
    at Object.define (vendor.js:17371)
    at __webpack_require__ (vendor.js:20)
    at Object.module (vendor.js:17297)
    at __webpack_require__ (vendor.js:20)
    at vendor.js:63
    at vendor.js:66
(anonymous) @ vendor.js:27659
define @ vendor.js:17367
define @ vendor.js:17371
__webpack_require__ @ vendor.js:20
module @ vendor.js:17297
__webpack_require__ @ vendor.js:20
(anonymous) @ vendor.js:63
(anonymous) @ vendor.js:66
```
The corresponding row is the last of
```js
if ( true ) {
	!(__WEBPACK_AMD_DEFINE_ARRAY__ = [], __WEBPACK_AMD_DEFINE_RESULT__ = (function() {
		return jQuery;
	}).apply(exports, __WEBPACK_AMD_DEFINE_ARRAY__),
				__WEBPACK_AMD_DEFINE_RESULT__ !== undefined && (module.exports = __WEBPACK_AMD_DEFINE_RESULT__));
}
```

Additionally installed packages/run commands:
```bash
apt -yqq install gnupg2 apt-transport-https
curl -sL https://deb.nodesource.com/setup_8.x | bash -
curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add -
echo "deb https://dl.yarnpkg.com/debian/ stable main" > /etc/apt/sources.list.d/yarn.list
apt -yqq update && apt -yqq install nodejs yarn
yarn install
yarn build
```

Did I missed anything?

**Comment by MyIgel** [comment]
The user-shifts page is broken -> the buttons in [user_shifts.html](https://github.com/engelsystem/engelsystem/blob/3c1bb5c8b468520e07b6cf57a652d3fbef598352/templates/user_shifts.html#L50) and [user_shifts.php](https://github.com/engelsystem/engelsystem/blob/3c1bb5c8b468520e07b6cf57a652d3fbef598352/includes/pages/user_shifts.php#L250) use `javascript:` links to trigger functions that are not available in the global scope

```
(unknown) Uncaught ReferenceError: setHours is not defined
    at <anonymous>:1:1
(anonymous)
```

**Comment by MyIgel** [comment]
After some testing it seems to work just fine -> @msquare could you test and merge it?

**Comment by msquare** [comment]
I found that the project is now too complicated to build for beginners, who only want to install the system. To merge this pull request we need an automated build that serves simple downloadable install files to extract and run. Then we need to split the readme into two sections: Simple install and download source, build and contribute.

Is this automated build possible on Github? Or should we switch to Gitlab?

Please also take a look on https://app.codacy.com/app/engelsystem/engelsystem/pullRequest?prid=1318180

**Comment by MyIgel** [comment]
Hm, that is a point -> I think we should use releases for that, that would allow us to easily publish zip files with all required files included. 
They can be also build with GitLab (integration request in #409 (Gitlab CI @chaos.expert)) so that they only need to be uploaded to GitHub after it; I can extend the build config to create artifacts with the whole pre build project

**Comment by msquare** [comment]
Yes, please!

**Comment by msquare** [comment]
@marudor: Please use only relative paths, the `/` at the beginning has to go.
```
@font-face {
	font-family: 'angelFont';
	src:url(/assets/96cc4a8f784e52ac0749c3104e9f4bde.eot);
[...]
}
```

**Comment by MyIgel** [comment]
That commit creates an archive file with all project sources included (except for git files and not needed cached files) so after deciding how we version the engelsystem (semantic versioning would be nice) we should create an 'legacy' release (maybe the version of the last congress?) and add Betas/RC's for the next improvements

**Comment by msquare** [comment]
What about codacy? We should decide if we use single or double quotes in javascript. Codacy is set to double quotes, but i can change this. At least I don't want to introduce >30 code style errors with this pull request. Please take a look @marudor.

Icons are fine now. Thanks.

@MyIgel: When this pull request is merged, we make a 1.0.0 release, providing the file from the build script as download.

**Comment by marudor** [comment]
JS usually uses single quotes.

**Pull request merged**